sips/sip-012/sip-012-cost-limits-network-upgrade.md

```diff
@@ -0,0 +1,674 @@
+
+The SIP will be considered Recommended if wallets indicating support for the upgrade (through either mechanism) add up to > 10% of circulating supply of STX (approx 120M).
+
+In terms of how these cost-limits would actually be applied, this SIP proposes the following:
+* Add new functionality to stacks-blockchain that uses the current cost-limits by default and uses new cost-limits if the burn block height exceeds a configurable parameter (could be a compile time configuration to avoid runtime issues)
+* Once a BTC block number has been determined, ship a new stacks-blockchain release at least one week before to give miners and node operators time to upgrade before the upgrade block height is reached
+* In the subsequent release, remove all usage of the old cost-limits and just use the new cost-limits by default
+
+### Default Cost Functions
+
+Based on results from the [clarity-benchmarking](https://github.com/blockstack/clarity-benchmarking) project, we propose
+new default cost functions. The new costs are supplied in the form of a new Clarity smart contract in [Appendix A](#appendix-a).
+
+### Block Limit Changes
+
+In addition to the runtime cost changes, we propose increasing the
+block limits for MARF reads and writes. Based on the expected performance
+improvements in the implementation of the MARF (see [issue #2869](https://github.com/blockstack/stacks-blockchain/issues/2869))
+we propose doubling the current limits on blocks:
+
+```rust
+pub const BLOCK_LIMIT_MAINNET: ExecutionCost = ExecutionCost {
+    write_length: 15_000_000, // roughly 15 mb
+    write_count: 15_500,
+    read_length: 100_000_000,
+    read_count: 15_500,
+    runtime: 5_000_000_000,
+};
+```
+
+### Changes to Static vs. Dynamic Tabulation of Costs
+
+The cost assessment in Clarity for most data-handling functions (e.g.,
+`map-get?`) use the static cost of the fetch rather than the dynamic
+cost. For more information, see [issue #2864](https://github.com/blockstack/stacks-blockchain/issues/2864)
+in the `stacks-blockchain` repository.
+
+There were two motivating reasons for doing this:
+
+* It makes static analysis of costs easier, because the cost assessed
+  at runtime would always use the declared size of the map entry.
+* It allows the cost to be assessed before running the operation.
+
+However, it's pretty easy to argue that these reasons aren't all that
+essential. For (1), static analysis is always going to overestimate
+anyways, so why not allow the real cost to match the actual runtime
+overhead more closely. For (2), allowing a single "speculative"
+evaluation before aborting a block due to cost overflow is not
+particularly burdensome to the network: the maximum size of an
+overread is a single Clarity value, 2MB.
+
+The benefit of using dynamic costs, however, could be great. Many
```

Comment:
This is a tautology -- benefits are always great.  "The benefit of using dynamic costs, however, is that many contracts that store potentially long lists to their data space would only be billed for the amount of space the list occupies, instead of its maximum size."